backend/cfg/cfg_regalloc_utils.ml

```diff
@@ -450,3 +455,83 @@ let may_use_stack_operands_everywhere :
   may_use_stack_operands_array map instr.arg;
   may_use_stack_operands_array map instr.res;
   All_spilled_registers_rewritten
+
+let insert_block :
+    Cfg_with_layout.t ->
+    Cfg.BasicInstructionList.t ->
+    after:Cfg.basic_block ->
+    next_instruction_id:(unit -> Instruction.id) ->
+    unit =
+ fun cfg_with_layout body ~after:predecessor_block ~next_instruction_id ->
+  let cfg = Cfg_with_layout.cfg cfg_with_layout in
+  let successors =
+    Cfg.successor_labels ~normal:true ~exn:false predecessor_block
+  in
+  (* CR gyorsh: this function is currently only when |successors|>=1 and |body|
+     > 0*)
+  if Label.Set.cardinal successors = 0
+  then
+    Misc.fatal_errorf
+      "Cannot insert a block after block %a: it has no successors" Label.print
+      predecessor_block.start;
+  let last_insn =
+    match Cfg.BasicInstructionList.last body with
+    | None -> Misc.fatal_error "Inserting an empty block"
+    | Some i -> i
+  in
+  let copy (i : Cfg.basic Cfg.instruction) : Cfg.basic Cfg.instruction =
+    { i with id = next_instruction_id () }
+  in
+  (* copy body if there is more than one successor *)
+  let first = ref true in
+  let get_body () =
+    if !first
+    then (
+      first := false;
+      body)
+    else
+      let new_body = Cfg.BasicInstructionList.make_empty () in
+      Cfg.BasicInstructionList.iter body ~f:(fun instr ->
+          Cfg.BasicInstructionList.add_end new_body (copy instr));
+      new_body
+  in
+  Label.Set.iter
+    (fun successor_label ->
+      let successor_block = Cfg.get_block_exn cfg successor_label in
+      let start = Cmm.new_label () in
+      let block : Cfg.basic_block =
+        { start;
+          body = get_body ();
+          terminator =
+            { (* The [successor_block] is the only successor. *)
+              desc = Cfg.Always successor_label;
+              arg = [||];
+              res = [||];
+              dbg = last_insn.dbg;
+              fdo = last_insn.fdo;
+              live = last_insn.live;
+              stack_offset = last_insn.stack_offset;
+              id = next_instruction_id ();
+              irc_work_list = Unknown_list
+            };
+          (* The [predecessor_block] is the only predecessor. *)
+          predecessors = Label.Set.singleton predecessor_block.start;
+          stack_offset = predecessor_block.terminator.stack_offset;
+          exn = None;
```

Comment:
If we insert _e.g._ between two blocks
with `Some e`, wouldn't it make sense
for the one in the middle to also have
`Some e`?

Reply:
I don't think so. If a block cannot raise, it has no exceptional successors, and basic instructions cannot raise in this PR.

Reply:
And neither can `Always`

Reply:
From my understand the `block.exn` is only for bookkeeping in the CFG and doesn't contribute in `cfg_to_linear` as the push and poptraps are explicit already.

And I believe this is the correct way to bookkeep according to other code.

Reply:
I mean that I believe that already in the `main`:
if there is a block that has `block.can_raise = false` but there are traps pushed on the stack then still `block.exn = None`.

So the `block.exn` isn't "top-most trap on the" stack but "where will we go if block raises.

But if the block doesn't raise then the `block.exn` is `None` regargless of traps being on stack or not.

So assuming there are no pushtraps or poptraps in the body of the added block and it cannot raise (which is true because only terminators can raise but `Always` can't) then the added block should have `block.exn = None`.

Reply:
I concur, and there is probably no
problems in practice. I am a bit
worried some parts of the code may
have used `exn` to know something
about the trap stack.

Reply:
Then I believe that code was already incorrect since couple months, because that's how `cfgize` generates CFG.

Reply:
Oh, interesting; that sounds right.
(Well "right" as in "true"...)

Reply:
If you take a look at `successor_labels` in `cfg.ml` for example, having `can_raise = false` and `exn = Some label` would be incorrect according to that function.

Reply:
> If you take a look at `successor_labels` in `cfg.ml` for example, having `can_raise = false` and `exn = Some label` would be incorrect according to that function.

it's correct, but overly conservative.